### PR TITLE
 Project 2: Partial Redundancy Elimination

### LINKED ISSUE #60: Project 2 Proposal: Partial Redundancy Elimination
**What will you do?**
I will implement a partial redundancy elimination pass. The goal is to eliminate expressions that are redundant on some but not necessarily all paths through a program.

**How will you do it?**
I found [this slides](http://www.cs.cmu.edu/afs/cs/academic/class/15745-s12/public/lectures/L10-PRE-1up.pdf) helpful, and it also lists a list of references.

The algorithm roughly contains:
* find partially available expressions.
* for each such expression, check if safety(doesn't change program behavior) and performance(doesn't add new computations) are guaranteed after we make this expression fully available. If so then we can make it fully available and eliminate it.

There are more details about how to implement this algorithm in the slides.

**How will you empirically measure success?**
To make sure my implementation is correct, I will create test cases to check:
* no non-safe expressions are eliminated, such as insertions that can cause an exception.
* no expressions are eliminated by introducing new computations in any path.
* the result after this optimization matches the original.

For the performance part, I will measure static instruction count change and wall-clock time speedup using test cases. I am also wondering if I can measure the numbers of instructions in the execution traces because it seems we don't have an existing tool.

I expect the instruction count to increase because of computation insertions. But there should be significant wall-clock time speedup.

**Team members:**
Just me.

### PR DESCRIPTION
**Pull request by Neroysq**
closes #60

### CODE REVIEW

File: content/blog/2019-10-23-brilpre.md [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+| ./test/pre_test/complex_loop3.json        | 24         | 26        | 8.3%  | 66004         | 66006        | 0.0%  | 63999              | 53001             | -17.2% |
+| ./test/pre_test/gcd.json                  | 16         | 16        | 0.0%  | 141           | 141          | 0.0%  | 92                 | 92                | 0.0%   |
+| ./test/pre_test/factorial.json            | 14         | 14        | 0.0%  | 100008        | 100008       | 0.0%  | 100003             | 100003            | 0.0%   |
+| ./test/df_test/cond.json                  | 15         | 15        | 0.0%  | 9             | 9            | 0.0%  | 3                  | 3                 | 0.0%   |
+| ./test/df_test/fact.json                  | 13         | 13        | 0.0%  | 62            | 62           | 0.0%  | 42                 | 42                | 0.0%   |
+
+Above is a table showing all the results, 
+unfortunately for most cases, the improvement is not significant, 
+part of the reasons should be that most programs are short and do not involve loops.
```
Comment (sampsyo) [inline content/blog/2019-10-23-brilpre.md]: Part of the reason *is*? Not just "should be."

File: content/blog/2019-10-23-brilpre.md [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+But for all programs with loops that I manually designed,
+this tool can successfully detect them, generate correct new code,
+and provide a significant performance improvement.
```
Comment (sampsyo) [inline content/blog/2019-10-23-brilpre.md]: It is cool that you ran your optimization on such a large number of programs! As you know, most of these were written as tiny tests to check the correctness of individual transformations, so they're not really realistic code in any sense. So I think it's fine that there were no optimization opportunities for most of them.

**Review by sampsyo — CHANGES_REQUESTED** [review]
Nice work! I especially enjoyed your very large evaluation with honest conclusions about what actually improves. I have a few small writing suggestions within.

File: content/blog/2019-10-23-brilpre.md [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+While preserving *computational optimality* (no partial redundant expression computations), 
+this algorithm guarantees *lifetime optimality* (computations are placed as early as necessary but as late as possible).
+
+This algorithm involves four passes over the control flow graph (CFG) of the program.
+
+Before the passes, this algorithm adds a new empty block between each block with multiple predecessors and each of its predecessors. And in the end, all empty blocks are removed. 
+
+#### Pass 1: Anticipated Expressions
+An expression is *anticipated* at a point if it is certain to be evaluated 
+along any path before defined (variables involved are reassigned).
```
[the remark this answers is not in the document]
Comment (Neroysq) [reply content/blog/2019-10-23-brilpre.md]: I rephrase it to "along any path before this expression's value is changed 
(any variables its evaluation depends on are reassigned)."

File: content/blog/2019-10-23-brilpre.md [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+I leveraged some code from my last project, such as the parsing of Bril JSON files.
+
+There are some tricks applied when implementing this algorithm:
+1. When building the control flow graph, 
+the implementation treats each instruction as one block. 
+So during all the passes, we can ensure that each block can only contain one instruction.
+This decision makes analyzing easier.  
+
+2. When inserting new variables and labels, we must make sure the new names are unique,
+the way I do is to find the smallest `n` where the string `n`*"_" is not a prefix of any
```
[the remark this answers is not in the document]
Comment (Neroysq) [reply content/blog/2019-10-23-brilpre.md]: Yes, you are definitely correct, but I just happened to do it that way.

File: content/blog/2019-10-23-brilpre.md [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+existing variables in the code. 
+Then I can create any name and put this prefix on to get rid of any naming conflict.
+
+3. When storing a value operation, this tool first normalizes the expression
+to try to make arguments sorted. If the order of the arguments need to be reversed, 
+this tool would also reverse the operation type (i.e., `le` to `ge`, `add` to `add`).  
+Then we can store expressions as strings in later steps.
```
[the remark this answers is not in the document]
Comment (Neroysq) [reply content/blog/2019-10-23-brilpre.md]: It is to maximize CSE opportunities - to make `a*b` and `b*a` equivalent.

File: content/blog/2019-10-23-brilpre.md [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+
+To measure lines of code, I wrote a script in Python to count lines of instructions in source code.
+
+To count instructions executed, 
+I hacked the reference interpreter `brili` to count instructions and show it at the end when executing.
+However, I found this number not representative: 
+Even PRE gets rid of redundant evaluations, it doesn't decrease the number of instructions executed,
+because it only replaces original evaluations with `id` operations instead of removing them.
+
+Therefore, I only count those computationally significant operations (all value operations except `id`),
```
[the remark this answers is not in the document]
Comment (Neroysq) [reply content/blog/2019-10-23-brilpre.md]: That also sounds ok to me. It should not affect the result too much.

File: content/blog/2019-10-23-brilpre.md (hunk 2 above)
Comment (Neroysq) [reply content/blog/2019-10-23-brilpre.md]: Glad that you think it is cool!

**Review by sampsyo — APPROVED** [review]
Looks great; thanks!!
